compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2267,6 +2270,16 @@ pub enum Rvalue<'tcx> {
     /// away after type-checking and before lowering.
     Aggregate(Box<AggregateKind<'tcx>>, Vec<Operand<'tcx>>),
 
+    /// Creates a value of the variant with all fields uninitialized
+    ///
+    /// Specifically, this has the property that once all fields of the variant are also
+    /// initialized, the memory of the place this is assigned to is a valid value for the type, and
+    /// that the `discriminant` rvalue for that value will yield the discriminant of the variant
+    /// here.
+    ///
+    /// Only allowed for ADTs.
+    UninitVariant(Ty<'tcx>, VariantIdx),
```

Comment:
I just realized there is something fundamentally sketch about this.

My idea of giving semantics to MIR is that every `Rvalue` (and in general, every expression in value position) evaluate to an element of an `enum` [like this](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/value-domain.md), and then when the value is stored to memory we use the value representation relation to find an encoding of this high-level value in low-level Abstract Machine Bytes. This relation also implicitly encodes the validity invariant.

However, that means that we *cannot have* an "uninit" (r)value of most types, since their representation relation would simply now allow such a value (i.e., would not relate it to any memory representation). This is essentially just as cursed as `mem::uninitialized`...

Reply:
Ah, yeah, so if this is the model we want then this idea indeed does not work. I have been operating under a different model, which is that `Rvalue`s and `Operand`s are always just a bag of bytes. Operations like dereferencing, `Rvalue::BinaryOp`, and such would then treat the bytes in some interesting way.

I can indeed turn this into a `StatementKind` then, if that fits better with existing models.

Reply:
In general, we certainly want assignment and argument parsing to treat values in the high-level way I described, so it's not just primitive operations. `let b: bool = unsafe { ptr.read() };` is explicitly intended to be insta-UB if the byte read is not 0 or 1.

So this would have to be a special case inside `StatementKind::Assign`, which seems odd.

Reply:
Well, we could still make that insta UB at the place to value conversion. The value that is produced does not have to be some high level type in order for `bool` to specify that place to value conversions at that type are UB if they do not yield `[0]` or `[1]`

Reply:
That seems conceptually rather awkward though -- it would mean `=` actually assigns a "raw sequence of bytes" but for some reason `Operand::{Copy,Move}` round-trip through a high-level value and back. This feels very bolted on.

I'd much rather define things such that `=` assigns a high-level value -- the right-hand side computes that value, then it is stored to the place denoted by the left-hand side.